stress: failed test in cockroach/kv/kv.test: TestEvictOnFirstRangeGossip

Binary: cockroach/static-tests.tar.gz sha: https://github.com/cockroachdb/cockroach/commits/2d7438d7883ac759b04c3b87de2ee46003a7b5ff

Stress build found a failed test:

```
=== RUN   TestEvictOnFirstRangeGossip
W160828 04:02:12.909873 gossip/gossip.go:1070  not connected to cluster; use --join to specify a connected node
--- FAIL: TestEvictOnFirstRangeGossip (0.00s)
	dist_sender_test.go:667: expected one first range lookup, got 2
```

Run Details:

```
0 runs so far, 0 failures, over 5s
0 runs so far, 0 failures, over 10s
0 runs so far, 0 failures, over 15s
0 runs so far, 0 failures, over 20s
0 runs so far, 0 failures, over 25s
16 runs so far, 0 failures, over 30s
16 runs so far, 0 failures, over 35s
16 runs so far, 0 failures, over 40s
16 runs so far, 0 failures, over 45s
16 runs so far, 0 failures, over 50s
17 runs completed, 1 failures, over 54s
FAIL
```
Please assign, take a look and update the issue accordingly.

--- a/kv/dist_sender_test.go
+++ b/kv/dist_sender_test.go
@@ -594,9 +594,6 @@ func makeGossip(t *testing.T, stopper *stop.Stopper) *gossip.Gossip {
 	if err := g.AddInfo(gossip.KeySentinel, nil, time.Hour); err != nil {
 		t.Fatal(err)
 	}
-	if err := g.AddInfoProto(gossip.KeyFirstRangeDescriptor, &testRangeDescriptor, time.Hour); err != nil {
-		t.Fatal(err)
-	}
 
 	return g
 }
@@ -634,7 +631,7 @@ func TestEvictOnFirstRangeGossip(t *testing.T) {
 	rDB := MockRangeDescriptorDB(func(key roachpb.RKey, _, _ bool) (
 		[]roachpb.RangeDescriptor, []roachpb.RangeDescriptor, *roachpb.Error,
 	) {
-		if len(key) == 0 {
+		if key.Equal(roachpb.KeyMin) {
 			atomic.AddInt32(&numFirstRange, 1)
 		}
 		return []roachpb.RangeDescriptor{desc}, nil, nil
@@ -650,10 +647,11 @@ func TestEvictOnFirstRangeGossip(t *testing.T) {
 	ds := NewDistSender(ctx, g)
 
 	anyKey := roachpb.Key("anything")
+	rAnyKey := keys.MustAddr(anyKey)
 
 	call := func() {
 		if _, _, err := ds.rangeCache.LookupRangeDescriptor(
-			context.Background(), keys.MustAddr(anyKey), nil, false, false,
+			context.Background(), rAnyKey, nil, false, false,
 		); err != nil {
 			t.Fatal(err)
 		}
@@ -661,6 +659,10 @@ func TestEvictOnFirstRangeGossip(t *testing.T) {
 
 	// Perform multiple calls and check that the first range is only looked up
 	// once, with subsequent calls hitting the cache.
+	//
+	// This potentially races with the cache-evicting gossip callback on the
+	// first range, so it is important that the first range descriptor's state
+	// in gossip is stable from this point forward.
 	for i := 0; i < 3; i++ {
 		call()
 		if num := atomic.LoadInt32(&numFirstRange); num != 1 {
@@ -759,6 +761,10 @@ func TestRetryOnWrongReplicaError(t *testing.T) {
 	defer stopper.Stop()
 
 	g := makeGossip(t, stopper)
+	if err := g.AddInfoProto(gossip.KeyFirstRangeDescriptor, &testRangeDescriptor, time.Hour); err != nil {
+		t.Fatal(err)
+	}
+
 	// Updated below, after it has first been returned.
 	badStartKey := roachpb.RKey("m")
 	newRangeDescriptor := testRangeDescriptor
@@ -832,6 +838,10 @@ func TestRetryOnWrongReplicaErrorWithSuggestion(t *testing.T) {
 	defer stopper.Stop()
 
 	g := makeGossip(t, stopper)
+	if err := g.AddInfoProto(gossip.KeyFirstRangeDescriptor, &testRangeDescriptor, time.Hour); err != nil {
+		t.Fatal(err)
+	}
+
 	// Updated below, after it has first been returned.
 	goodRangeDescriptor := testRangeDescriptor
 	badRangeDescriptor := testRangeDescriptor

--- a/kv/range_cache.go
+++ b/kv/range_cache.go
@@ -215,7 +215,7 @@ type evictionToken struct {
 	doOnce    sync.Once                                 // assures that do and doReplace are run up to once.
 	doLocker  sync.Locker                               // protects do and doReplace.
 	do        func() error                              // called on eviction.
-	doReplace func(rs ...roachpb.RangeDescriptor) error // called after eviction on evictAndReplace.
+	doReplace func(rs ...roachpb.RangeDescriptor) error // called after eviction on EvictAndReplace.
 }
 
 func (rdc *rangeDescriptorCache) makeEvictionToken(prevDesc *roachpb.RangeDescriptor, evict func() error) *evictionToken {

